feat: Add audio synchronization with timestamp support

isaiahdahl opened the pull request:
Implements comprehensive audio synchronization for ChordSheetJS with timestamp directives and a framework-agnostic TimestampManager.

Features:
- New {timestamp:} directive in ChordPro format with {tm:} shorthand
- Support for pipe-delimited timestamps for repeated sections
- Line-level and inline timestamp placement
- MeasuredHtmlFormatter auto-height mode (pageSize.height: 'auto')
- Framework-agnostic TimestampManager class for audio sync
- Auto-scrolling with configurable offset and behavior
- Active element highlighting
- Playground AudioSyncPanel for testing

Technical changes:
- Added timestamps property to Line and ChordLyricsPair
- Enhanced SongBuilder to parse timestamp tags
- Modified layout engine to propagate timestamps through layout calculation
- Updated PositionedHtmlRenderer to add data-timestamp attributes
- Fixed Song.linesToParagraphs() to preserve timestamps from non-renderable lines
- Fixed Line.clone() and SongMapper to preserve timestamps during transformations
- Added comprehensive test coverage (12+ new tests)

Fixes:
- Timestamps now properly preserved through key changes, capo, and transposition
- Non-renderable lines with timestamps correctly attached to next renderable line
- ChordLyricsPair splitting preserves timestamps

Breaking changes: None - all changes are backward compatible

isaiahdahl committed:
feat: Add audio synchronization with timestamp support

Implements comprehensive audio synchronization for ChordSheetJS with timestamp directives
and a framework-agnostic TimestampManager.

Features:
- New {timestamp:} directive in ChordPro format with {tm:} shorthand
- Support for pipe-delimited timestamps for repeated sections
- Line-level and inline timestamp placement
- MeasuredHtmlFormatter auto-height mode (pageSize.height: 'auto')
- Framework-agnostic TimestampManager class for audio sync
- Auto-scrolling with configurable offset and behavior
- Active element highlighting
- Playground AudioSyncPanel for testing

Technical changes:
- Added timestamps property to Line and ChordLyricsPair
- Enhanced SongBuilder to parse timestamp tags
- Modified layout engine to propagate timestamps through layout calculation
- Updated PositionedHtmlRenderer to add data-timestamp attributes
- Fixed Song.linesToParagraphs() to preserve timestamps from non-renderable lines
- Fixed Line.clone() and SongMapper to preserve timestamps during transformations
- Added comprehensive test coverage (12+ new tests)

Fixes:
- Timestamps now properly preserved through key changes, capo, and transposition
- Non-renderable lines with timestamps correctly attached to next renderable line
- ChordLyricsPair splitting preserves timestamps

Breaking changes: None - all changes are backward compatible

isaiahdahl committed:
fix: ci and small adjustments still wip

martijnversluis commented on src/chord_sheet/song.ts:
```diff
@@ -419,39 +442,10 @@ Or set the song key before changing key:
   }
 
-  /**
-   * Returns a copy of the song with the directive value(s) set to the specified value(s).
```
Did you mean to cut down on comment size? I'd vote keeping the code samples in there, they end up in the rendered API docs as examples.

martijnversluis commented on test/chord_sheet/change_tempo.test.ts:
```diff
@@ -0,0 +1,187 @@
+
+  describe('with ChordPro parsing', () => {
+    it('scales timestamps when slowing down (120 BPM → 60 BPM)', () => {
+      const chordSheet = `{timestamp: 0:10}
```
You can use the `heredoc` template function and indent the contents:

```ts
      const chordSheet = heredoc`
        {timestamp: 0:10}
        [C]Hello [G]World
        {timestamp: 0:20}
        [Am]Second [F]line
      `;
```

martijnversluis reviewed (commented):
Wow you crunched out quite the feature! Also well tested 👍

isaiahdahl committed:
docs: restore API documentation in song.ts per review feedback

isaiahdahl committed:
refactor: extract SimpleEventEmitter class from TimestampManager

isaiahdahl committed:
style: use heredoc for multiline strings in change_tempo.test.ts

isaiahdahl committed:
feat(playground): collapse audio sync panel by default

isaiahdahl committed:
chore: bump max-lines limit to accommodate timestamp feature

isaiahdahl committed:
feat(audio-sync): enhance AudioSyncPanel with simulated playback and timestamp management

- Introduced simulated playback functionality to the AudioSyncPanel, allowing for audio synchronization without an audio file.
- Added logic to manage timestamps effectively, including visibility updates based on the presence of timestamps.
- Improved handling of maximum duration calculations based on timestamps, ensuring accurate playback behavior.
- Updated UI elements to reflect changes in playback state and enabled play/pause button functionality.
- Enhanced cleanup process to reset timestamp states when the panel is disposed.

This update builds on the existing timestamp features, providing a more robust audio sync testing experience.

isaiahdahl commented on src/playback/timestamp_manager.ts:
```diff
@@ -0,0 +1,521 @@
+  /**
+   * Register an event handler
+   */
+  private on(event: string, callback: EventCallback): () => void {
```
Not entirely sure what exactly you meant by extracting it out but interrupted it as creating a SimpleEventEmitter class and having the TimestampManager extend that. 

Thoughts?

martijnversluis commented:
I'm not sure if I missed this when reviewing, but I just realised the `TimestampManager` does "browser things", while (so far) the entire library was strictly platform agnostic. Any way this can be moved to either the playground or the editor?
